src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -949,6 +967,26 @@ - (NSArray *)_getAttributesToReportWithReportedValues:(NSArray<NSDictionary<NSSt
                     MTR_LOG_INFO("%@ report %@ value filtered - same values as cache", self, attributePath);
                 }
             }
+
+            // If General Diagnostics UpTime attribute, save the likely better estimated start time. At OnSubscriptionEstablished
+            // time, this estimate will be used if it's more accurate. Note that this needs to be done because the current SDK
+            // implementation reports events after attributes, and the system startup event resets the estimated start time.
+            if ((attributePath.cluster.unsignedLongValue == MTRClusterGeneralDiagnosticsID)
+                && (attributePath.attribute.unsignedLongValue == MTRClusterGeneralDiagnosticsAttributeUpTimeID)) {
+                // verify that the uptime is indeed the data type we want
+                if ([attributeDataValue[MTRTypeKey] isEqual:MTRUnsignedIntegerValueType]) {
+                    NSNumber * upTimeNumber = attributeDataValue[MTRValueKey];
+                    NSTimeInterval upTime = upTimeNumber.unsignedLongLongValue; // UpTime unit is defined as seconds in the spec
+                    NSDate * oldSystemStartTime = _estimatedStartTimeFromGeneralDiagnosticsUpTime;
+                    NSDate * potentialSystemStartTime = [NSDate dateWithTimeIntervalSinceNow:-upTime];
+                    if (!_estimatedStartTimeFromGeneralDiagnosticsUpTime
+                        || ([potentialSystemStartTime compare:_estimatedStartTime] == NSOrderedAscending)) {
```

Comment:
Why are we checking _estimatedStartTimeFromGeneralDiagnosticsUpTime but then if it's not null comparing to _estimatedStartTime?

Reply:
Thank you for catching this - it should be comparing with `_estimatedStartTimeFromGeneralDiagnosticsUpTime` here. I'll fix this in a moment.